# Learn PyTorch

Step-by-step mastering of PyTorch:

http://pytorch.org/tutorials/beginner/blitz/tensor_tutorial.html#sphx-glr-beginner-blitz-tensor-tutorial-py



In [1]:
# As usual, a bit of setup
from __future__ import print_function
import time
import numpy as np
import matplotlib.pyplot as plt
import platform
import torch

%matplotlib inline
plt.rcParams['figure.figsize'] = (10.0, 8.0) # set default size of plots
plt.rcParams['image.interpolation'] = 'nearest'
plt.rcParams['image.cmap'] = 'gray'

# for auto-reloading external modules
# see http://stackoverflow.com/questions/1907993/autoreload-of-modules-in-ipython
%load_ext autoreload
%autoreload 2

print("Python version: ", platform.python_version())

Python version:  3.6.3


## Some basics

In [6]:
x = torch.Tensor(5, 3)   # uninitialized
print(x)

x = torch.rand(5, 3)  # randomized
print(x)

print(x.size())  # torch.Size is a tuple

y = torch.rand(5, 3)
print(y)

# Three ways to add Tensors

print(x + y)  # Way 1

print(torch.add(x, y))  # Way 2

result = torch.Tensor(5, 3)  # Way 3
torch.add(x, y, out=result)
print(result)


-1.0030e+17  4.5642e-41 -1.0030e+17
 4.5642e-41  0.0000e+00  0.0000e+00
 0.0000e+00  0.0000e+00  0.0000e+00
 0.0000e+00  0.0000e+00  0.0000e+00
 8.9683e-44  0.0000e+00  1.5695e-43
[torch.FloatTensor of size 5x3]


 0.3232  0.6961  0.7682
 0.0437  0.1695  0.4603
 0.0755  0.0513  0.5006
 0.7241  0.9199  0.4430
 0.6940  0.1012  0.7336
[torch.FloatTensor of size 5x3]

torch.Size([5, 3])

 0.0650  0.4111  0.9006
 0.3754  0.6471  0.6847
 0.3545  0.0664  0.0459
 0.4200  0.3889  0.0885
 0.3900  0.2364  0.9111
[torch.FloatTensor of size 5x3]


 0.3882  1.1072  1.6688
 0.4191  0.8166  1.1450
 0.4300  0.1176  0.5465
 1.1441  1.3089  0.5315
 1.0840  0.3375  1.6447
[torch.FloatTensor of size 5x3]


 0.3882  1.1072  1.6688
 0.4191  0.8166  1.1450
 0.4300  0.1176  0.5465
 1.1441  1.3089  0.5315
 1.0840  0.3375  1.6447
[torch.FloatTensor of size 5x3]


 0.3882  1.1072  1.6688
 0.4191  0.8166  1.1450
 0.4300  0.1176  0.5465
 1.1441  1.3089  0.5315
 1.0840  0.3375  1.6447
[torch.FloatTensor of size 5x3

## In place operation

Any operation that mutates a tensor in-place is post-fixed with an _. For example: x.copy_(y), x.t_(), will change x.


In [7]:
print(y)
y.add_(x)
print(y)


 0.0650  0.4111  0.9006
 0.3754  0.6471  0.6847
 0.3545  0.0664  0.0459
 0.4200  0.3889  0.0885
 0.3900  0.2364  0.9111
[torch.FloatTensor of size 5x3]


 0.3882  1.1072  1.6688
 0.4191  0.8166  1.1450
 0.4300  0.1176  0.5465
 1.1441  1.3089  0.5315
 1.0840  0.3375  1.6447
[torch.FloatTensor of size 5x3]



## Indexing, Reshape, etc.

Indexing and slicing is similar to Numpy.

Resizing: If you want to resize/reshape tensor, you can use torch.view


In [10]:
print(x)
print(x[:, 1])  # indexing, slicing

x = torch.randn(4, 4)
y = x.view(16)
print(y)
z = x.view(-1, 8)  # the size -1 is inferred from other dimensions
print(z)
print(x.size(), y.size(), z.size())


 1.0060  0.7897  0.1215  1.4496
-0.6373 -0.0289 -0.0835 -0.2037
-0.8296 -1.1209  2.0529  1.3891
 1.1334 -0.1597  0.0246  2.0013
[torch.FloatTensor of size 4x4]


 0.7897
-0.0289
-1.1209
-0.1597
[torch.FloatTensor of size 4]


-1.2132
-1.2730
-1.0381
-0.8865
 0.0797
-0.2809
-0.8449
 0.4542
-1.4533
-0.7981
-1.0626
 0.8023
 1.1387
 0.7386
 0.8643
 0.9609
[torch.FloatTensor of size 16]


-1.2132 -1.2730 -1.0381 -0.8865  0.0797 -0.2809 -0.8449  0.4542
-1.4533 -0.7981 -1.0626  0.8023  1.1387  0.7386  0.8643  0.9609
[torch.FloatTensor of size 2x8]

torch.Size([4, 4]) torch.Size([16]) torch.Size([2, 8])


## Numpy Bridge

The Torch Tensor and NumPy array will share their underlying memory locations, and changing one will change the other.

In [30]:
# Converting a Torch Tensor to a NumPy Array

a = torch.ones(5)
print(a)
print(a.size())

W = np.random.randn(3,5)

b = a.numpy()   # 
print(b.shape)  
print(np.dot(W,b))

c = a.numpy().reshape(-1,1)  
print(c.shape)
print(np.dot(W,c))

a.add_(2)   # Note that changing the Tensor a will change the numpy arrays b and c
print (b)
print (c)



 1
 1
 1
 1
 1
[torch.FloatTensor of size 5]

torch.Size([5])
(5,)
[ 2.00629776 -0.11403814 -3.50661283]
(5, 1)
[[ 2.00629776]
 [-0.11403814]
 [-3.50661283]]
[ 3.  3.  3.  3.  3.]
[[ 3.]
 [ 3.]
 [ 3.]
 [ 3.]
 [ 3.]]


In [32]:
# Converting NumPy Array to Torch Tensor

a = np.ones(5)
b = torch.from_numpy(a)
print(b)
np.add(a, 1, out=a)  # changing numpy a will change tensor b
print(a)
print(b)


 1
 1
 1
 1
 1
[torch.DoubleTensor of size 5]

[ 2.  2.  2.  2.  2.]

 2
 2
 2
 2
 2
[torch.DoubleTensor of size 5]



## Cuda

Tensors can be moved onto GPU using the .cuda method.

In [34]:
# let us run this cell only if CUDA is available
if torch.cuda.is_available():
    x = x.cuda()
    y = y.cuda()
    x + y
    
print (torch.cuda.is_available())

False


## Autograd: automatic differentiation

The autograd package provides automatic differentiation for all operations on Tensors. It is a define-by-run framework, which means that your backprop is defined by how your code is run, and that every single iteration can be different.

http://pytorch.org/tutorials/beginner/blitz/autograd_tutorial.html

autograd.Variable is the central class of the package. It wraps a Tensor, and supports nearly all of operations defined on it. Once you finish your computation you can call .backward() and have all the gradients computed automatically.

You can access the raw tensor through the .data attribute, while the gradient w.r.t. this variable is accumulated into .grad.


In [38]:
import torch
from torch.autograd import Variable

x = Variable(torch.ones(2, 2), requires_grad=True)
print(x)

y = x + 2
print(y)

# y was created as a result of an operation, so it has a grad_fn.
print(y.grad_fn)

z = y * y * 3
out = z.mean()

print(z, out)

Variable containing:
 1  1
 1  1
[torch.FloatTensor of size 2x2]

Variable containing:
 3  3
 3  3
[torch.FloatTensor of size 2x2]

Variable containing:
 27  27
 27  27
[torch.FloatTensor of size 2x2]
 Variable containing:
 27
[torch.FloatTensor of size 1]



## Gradient and Function

There’s one more class which is very important for autograd implementation - a Function.

Variable and Function are interconnected and build up an acyclic graph, that encodes a complete history of computation. Each variable has a .grad_fn attribute that references a Function that has created the Variable (except for Variables created by the user - their grad_fn is None).

If you want to compute the derivatives, you can call .backward() on a Variable. If Variable is a scalar (i.e. it holds a one element data), you don’t need to specify any arguments to backward(), however if it has more elements, you need to specify a grad_output argument that is a tensor of matching shape.


In [41]:
# y was created as a result of an operation, so it has a grad_fn.
print(out.grad_fn)  # out a Variable created by z.mean()

out.backward()  # out.backward() is equivalent to doing out.backward(torch.Tensor([1.0]))

print(x.grad)


Variable containing:
 4.5000  4.5000
 4.5000  4.5000
[torch.FloatTensor of size 2x2]



** What happened up there??? **

We have that 
$$ o = {1\over4} {∑ z_{i}} $$
$$ z_{i} = 3{(x_{i}+2)}^2 $$

Therefore, 
$$ {∂o \over ∂x_{i}} = {{3 \over 2} {(x_{i}+2)}}$$

When $ x_{i} = 1 $, $ {∂o \over ∂x_{i}} = 4.5 $

In [54]:
x = torch.randn(3)
x = Variable(x, requires_grad=True)

y = x * 2
print (y)

while y.data.norm() < 1000:
    print (y.data.norm())
    y = y * 2

print(y)

gradients = torch.FloatTensor([0.1, 1.0, 0.0001])
y.backward(gradients)

print(x.grad)

Variable containing:
 1.9765
 1.4325
 1.6437
[torch.FloatTensor of size 3]

2.9428786269703626
5.885757253940725
11.77151450788145
23.5430290157629
47.0860580315258
94.1721160630516
188.3442321261032
376.6884642522064
753.3769285044128
Variable containing:
 1011.9902
  733.4418
  841.5737
[torch.FloatTensor of size 3]

Variable containing:
  102.4000
 1024.0000
    0.1024
[torch.FloatTensor of size 3]



** What happened up there? **


We have that 
$$ y = 2x $$
$$ y = y * 2^9 = x * 2^{10} $$

Therefore, 
$$ {∂y \over ∂x} = 1024$$

When upstream gradient is [0.1, 1.0, 0.0001]

$$ {∂f \over ∂x} = {∂f \over ∂y}{∂y \over ∂x} = \begin{pmatrix}
0.1\\
1.0\\
0.0001
\end{pmatrix} 1024 = \begin{pmatrix}
102.4\\
1024.0\\
0.1024
\end{pmatrix}$$

The ability to inject gradient into a computation graph is important for policy gradient implementation.  